util/reflect.go

```diff
@@ -367,61 +413,368 @@ type NodeInfo struct {
+	if schema.Key == "" {
+		return nil, nil, fmt.Errorf("getNodesList: path %v cannot traverse unkeyed list type %T", path, root)
+	}
+	if !IsValueMap(rv) {
+		// Only keyed lists can be traversed with a path.
+		return nil, nil, fmt.Errorf("getNodesList: root has type %T, expect map", root)
+	}
+	emptyKey := false
+	if len(path.GetElem()[0].GetKey()) == 0 {
+		DbgPrint("path %v at %T points to list with empty wildcard key", path, root)
+		emptyKey = true
+	}
+
+	listElementType := rv.Type().Elem().Elem()
+	listKeyType := rv.Type().Key()
+
+	var matchNodes []interface{}
+	var matchSchemas []*yang.Entry
+
+	// Iterate through all the map keys to see if any match the path.
+	for _, k := range rv.MapKeys() {
+		ev := rv.MapIndex(k)
+		DbgPrint("checking key %v, value %v", k.Interface(), ValueStr(ev.Interface()))
+		match := true
+		if !emptyKey { // empty key matches everything.
+			if !IsValueStruct(k) {
+				// Compare just the single value of the key represented as a string.
+				pathKey, ok := path.GetElem()[0].GetKey()[schema.Key]
+				if !ok {
+					return nil, nil, fmt.Errorf("gnmi path %v does not contain a map entry for the schema key field name %s, parent type %T",
+						path, schema.Key, root)
+				}
+				kv, err := getKeyValue(ev.Elem(), schema.Key)
+				if err != nil {
+					return nil, nil, err
+				}
+				match = (fmt.Sprint(kv) == pathKey)
+				DbgPrint("check simple key value (%s): %t", pathKey, match)
+			} else {
+				// Must compare all the key fields.
+				for i := 0; i < k.NumField(); i++ {
+					kfn := listKeyType.Field(i).Name
+					fv := ev.Elem().FieldByName(kfn)
+					if !fv.IsValid() {
+						return nil, nil, fmt.Errorf("element struct type %s does not contain key field %s", k.Type(), kfn)
+					}
+					nv := fv
+					if fv.Type().Kind() == reflect.Ptr {
+						// Ptr values are deferenced in key struct.
+						nv = nv.Elem()
+					}
+					kf, ok := listElementType.FieldByName(kfn)
+					if !ok {
+						return nil, nil, fmt.Errorf("element struct type %s does not contain key field %s", k.Type(), kfn)
+					}
+					pathKey, ok := path.GetElem()[0].GetKey()[pathStructTagKey(kf)]
+					if !ok {
+						// If the key is not filled, it is assumed to match.
+						continue
+					}
+					if pathKey != fmt.Sprint(k.Field(i).Interface()) {
+						match = false
+						break
+					}
+					DbgPrint("key field value %s matches", pathKey)
+				}
+			}
+		}
+
+		if match {
+			// Pass in the list schema, but the actual selected element
+			// rather than the whole list.
+			DbgPrint("key matches")
+			n, s, err := getNodesInternal(schema, ev.Interface(), popGNMIPath(path))
+			if err != nil {
+				return nil, nil, err
+			}
+			if n != nil {
+				matchNodes = append(matchNodes, n...)
+				matchSchemas = append(matchSchemas, s...)
+			}
+		}
+	}
+
+	if len(matchNodes) == 0 {
+		return nil, nil, nil
+	}
+	return matchNodes, matchSchemas, nil
+}
+
+// pathStructTagKey returns the string label of the struct field sf when it is
```

Comment:
This is only true when compression is enabled -- this code is robust to both cases, but the comment isn't entirely true.

Reply:
If path compression not enabled then I think the element that would've been compressed (e.g. config/name) cannot be part of the key anymore because it's now a container or list. In that case pathToSchema would just return "name" (if that is part of the key). I may be missing your point - let's talk about it live.

Reply:
Sorry, yes this was a nit that I was just pointing out. Feel free to ignore.